### ISSUE TITLE
Cargo clone_on_copy false positive or bad suggestion

### ISSUE THREAD

**Issue by alan-signal** [rust-lang/rust-clippy#7414]
Lint name: clone_on_copy

I tried this code:

```rust
    #[test]
    fn copy_on_clone_false_positive() {
        let mut empty = [];
        let mut array = [0, 1];
        assert_eq!(
            (&mut empty[..], &mut array.clone()[..]),
            array.split_at_mut(0)
        );
    }
```

```
$ cargo clippy --all-targets
```

I expected to see this happen: No warning.

Instead, this happened: This warning:

```
warning: using `clone` on type `[i32; 2]` which implements the `Copy` trait
   --> src/file.rs:470:35
    |
470 |             (&mut empty[..], &mut array.clone()[..]),
    |                                   ^^^^^^^^^^^^^ help: try removing the `clone` call: `array`
    |
    = note: `#[warn(clippy::clone_on_copy)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_copy

warning: 1 warning emitted

```

When I apply the suggested fix:

```rust
    #[test]
    fn copy_on_clone_false_positive() {
        let mut empty = [];
        let mut array = [0, 1];
        assert_eq!(
            (&mut empty[..], &mut array[..]),
            array.split_at_mut(0)
        );
    }
```

It no longer compiles:

```
error[E0499]: cannot borrow `array` as mutable more than once at a time
   --> src/file.rs:469:54
    |
469 |         assert_eq!((&mut empty[..], &mut array[..]), array.split_at_mut(0));
    |         ---------------------------------------------^^^^^------------------
    |         |                                |           |
    |         |                                |           second mutable borrow occurs here
    |         |                                first mutable borrow occurs here
    |         first borrow later used here
```

### Meta

- `cargo clippy -V`: `clippy 0.1.52 (9bc8c42b 2021-05-09)`
- `rustc -Vv`:
  ```
rustc 1.52.1 (9bc8c42bb 2021-05-09)
binary: rustc
commit-hash: 9bc8c42bb2f19e745a63f3445f1ac248fb015e53
commit-date: 2021-05-09
host: x86_64-apple-darwin
release: 1.52.1
LLVM version: 12.0.0
  ```

**Comment by antonilol** [comment]
I probably have a duplicate of this:

src/main.rs:
```rust
fn main() {
    let mut x = (123, 456);
    let _y = (&x.clone().0, &mut x);
}
```

Run `cargo clippy --fix --allow-no-vcs`:
```
    Checking a v0.1.0 (/tmp/a)
warning: failed to automatically apply fixes suggested by rustc to crate `a`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0502]: cannot borrow `x` as mutable because it is also borrowed as immutable
 --> src/main.rs:3:21
  |
3 |     let _y = (&x.0, &mut x);
  |              -------^^^^^^-
  |              ||     |
  |              ||     mutable borrow occurs here
  |              |immutable borrow occurs here
  |              immutable borrow later used here

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0502`.
Original diagnostics will follow.

warning: using `clone` on type `(i32, i32)` which implements the `Copy` trait
 --> src/main.rs:3:16
  |
3 |     let _y = (&x.clone().0, &mut x);
  |                ^^^^^^^^^ help: try removing the `clone` call: `x`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#clone_on_copy
  = note: `#[warn(clippy::clone_on_copy)]` on by default

warning: `a` (bin "a") generated 1 warning (run `cargo clippy --fix --bin "a" -p a -- --no-deps` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.07s
```

The problem is that the suggestion just removes the clone instead of replacing the clone with a copy.

A correct suggestion would be changing `&x.clone().0` to `&{ x }.0`. `x` is then correctly copied and then borrowed (in my case a function expected a borrowed value).

The not yet stable [`std::mem::copy`](https://doc.rust-lang.org/std/mem/fn.copy.html) could also be used when/if it stabilizes.
